src/Dialect/ONNX/Rewrite.cpp

```diff
@@ -282,6 +282,95 @@ class BinaryOpBroadcastAxisPattern : public OpRewritePattern<OP_TYPE> {
   }
 };
 
+// A pattern to turn
+//   `BinaryOp(Constant_X), ExpandOp(Constant_Y))`
+// into
+//   `ExpandOp(BinaryOp(Constant_X, Constant_Y))`
+// which put constants together so that BinaryOp can be folded. This pattern
+// only handles the case where one of the operand is a scalar constant. For such
+// a case, we can easily infer the shape operand for the resulting ExpandOp.
+
+template <typename OP_TYPE>
+Value EmitBinaryOp(
+    MultiDialectBuilder<OnnxBuilder> &create, Value lhs, Value rhs) {
+  llvm_unreachable("Unsupported operator");
+}
+
+template <>
+Value EmitBinaryOp<ONNXAddOp>(
+    MultiDialectBuilder<OnnxBuilder> &create, Value lhs, Value rhs) {
+  return create.onnx.add(lhs, rhs);
+}
+
+template <>
+Value EmitBinaryOp<ONNXDivOp>(
+    MultiDialectBuilder<OnnxBuilder> &create, Value lhs, Value rhs) {
+  return create.onnx.div(lhs, rhs);
+}
+
+template <>
+Value EmitBinaryOp<ONNXMulOp>(
+    MultiDialectBuilder<OnnxBuilder> &create, Value lhs, Value rhs) {
+  return create.onnx.mul(lhs, rhs);
+}
+
+template <>
+Value EmitBinaryOp<ONNXSubOp>(
+    MultiDialectBuilder<OnnxBuilder> &create, Value lhs, Value rhs) {
+  return create.onnx.sub(lhs, rhs);
+}
+
+template <typename OP_TYPE>
+class PropagateScalarConstantExpandPattern : public OpRewritePattern<OP_TYPE> {
+public:
+  using OpRewritePattern<OP_TYPE>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      OP_TYPE binaryOp, PatternRewriter &rewriter) const override {
+    Operation *op = binaryOp.getOperation();
+    Location loc = binaryOp.getLoc();
+
+    assert(op->getNumOperands() == 2 && "op must be binary");
+    Value lhs = op->getOperand(0);
+    Value rhs = op->getOperand(1);
+    Type outputType = op->getResult(0).getType();
+
+    // Match
+    //  - lhs is a scalar constant, and
+    //  - rhs is ExpandOp whose input is a scalar constant, or vice versa.
+    Value expandShape = nullptr;
+    auto matchValue = [&expandShape](Value v) -> Value {
+      Value res = v;
+      if (definedBy<ONNXExpandOp>(res)) {
+        auto expandOp = cast<ONNXExpandOp>(res.getDefiningOp());
+        res = expandOp.getInput();
+        expandShape = expandOp.getShape();
+      }
+      if (isDenseONNXConstant(res) && isScalarTensor(res))
+        return res;
+      return nullptr;
+    };
+    Value lhsConstant = matchValue(lhs);
+    Value rhsConstant = matchValue(rhs);
+    if (!expandShape || !lhsConstant || !rhsConstant)
+      return failure();
+    // Does not handle empty shape in ExpandOp, e.g. of type tensor<0xdtype>.
+    if (!hasShapeAndRank(expandShape))
+      return failure();
+    ArrayRef<int64_t> dims = getShape(expandShape.getType());
+    if ((dims.size() == 1) && (dims[0] == 0))
+      return failure();
+
+    // Rewrite
+    MultiDialectBuilder<OnnxBuilder> create(rewriter, loc);
+    Value res = create.onnx.expand(outputType,
+        EmitBinaryOp<OP_TYPE>(create, lhsConstant, rhsConstant), expandShape);
```

Comment:
if you clone binaryOp then you don't need EmitBinaryOp:
```
auto clone = rewriter.cloneWithoutRegions<OP_TYPE>(*op);
clone->setOperands({lhsConstant, rhsConstant});
clone.inferShapes([](mlir::Region &region) {}));
```
or
```
auto clone = create.onnx.createOpAndInferShapes<OP_TYPE>(lhsConstant, rhsConstant);
```
or
```
auto clone = rewriter.create<OP_TYPE>(lhsConstant, rhsConstant);
```
and rely on that all your ops have a builder that infers the result type (custom_builder_broadcast_to_same_type_ops_list in gen_onnx_mlir.py)

Reply:
I use `auto clone = create.onnx.createOpAndInferShapes<OP_TYPE>(lhsConstant, rhsConstant);`. Thanks!